webharness/harness.js

```diff
@@ -145,33 +215,34 @@ function run_next_test() {
     finish();
     return;
   }
-
-  var path = tests[current_test].path;
-  try {
-    current_window = window.open("/tests/" + path);
-  } catch(ex) {
-    harness_error(ex);
-    return;
-  }
-  current_window.onerror = test_error;
-  current_window.addEventListener("load", function() {
-    dump("loaded " + path + "\n");
-    send_message({"action": "test_loaded", "test": path});
-    // Wait for other side to have loaded this test.
-    wait_for_message().then(function (m) {
-      if (m.action != "test_loaded") {
-        //XXX: should this be fatal?
-        harness_error(new Error("Looking for test_loaded, got: " + JSON.stringify(m)));
-        return;
-      }
-      if (m.test != path) {
-        harness_error(new Error("Wrong test loaded on other side: " + JSON.stringify(m.test)));
-        return;
-      }
-      current_window.run_test(is_initiator);
+  var room_ready = connect_socket("test");
+  room_ready.then(()=>{
+    var path = tests[current_test].path;
+    try {
+      current_window = window.open("/tests/" + path);
+    } catch(ex) {
+      harness_error(ex);
+      return;
+    }
+    current_window.onerror = test_error;
+    current_window.addEventListener("load", function() {
+      dump("loaded " + path + "\n");
+      send_admin_message({"action": "test_loaded", "test": path});
+      // Wait for other side to have loaded this test.
+      wait_for_admin_message().then(function (m) {
+        if (m.action != "test_loaded") {
+          //XXX: should this be fatal?
+          harness_error(new Error("Looking for test_loaded, got: " + JSON.stringify(m)));
+          return;
+        }
+        if (m.test != path) {
+          harness_error(new Error("Wrong test loaded on other side: " + JSON.stringify(m.test)));
+          return;
+        }
+        current_window.run_test(is_initiator);
+      });
     });
   });
-  //TODO: timeout handling
```

Comment:
Can we please preserve this comment?

BTW a Promise race would now probably an easy way to implement timeouts here.